app/controllers/saved_pages_controller.rb

```diff
@@ -0,0 +1,72 @@
+class SavedPagesController < ApplicationController
+  include AuthenticatedApiConcern
+
+  before_action :check_saved_page_exists, only: %i[destroy show]
+
+  # GET /api/saved_pages
+  def index
+    @saved_pages = SavedPage.where(oidc_user_id: user.id)
+    render_api_response(saved_pages: present_pages(@saved_pages))
+  end
+
+  # GET /api/saved_pages/:page_path
+  def show
+    render_api_response(saved_page: present_page(saved_page))
+  end
+
+  # POST /api/saved_pages
+  def create
+    saved_page = SavedPage.new(oidc_user_id: user.id, page_path: params[:page_path])
+
+    if saved_page.save
+      render_api_response(saved_page: present_page(saved_page))
+    else
+      handle_create_saved_page_errors(saved_page)
+    end
+  end
+
+  # DELETE /api/saved_pages
+  def destroy
+    return head :no_content if saved_page.destroy!
+
+    raise ApiError::SavedPageNotDestroyed, { page_path: params[:page_path] }
+  end
+
+private
+
+  def handle_create_saved_page_errors(saved_page)
+    error_types = saved_page.errors.map(&:type)
+
+    if error_types.include?(:taken)
+      raise ApiError::SavedPageDuplicate, { page_path: params[:page_path] }
+    end
+
+    if error_types.include?(:blank)
+      raise ApiError::SavedPagePathRequired, { page_path: params[:page_path] }
+    end
+
+    if error_types.any?
+      raise ApiError::SavedPageNotPersisted, { page_path: params[:page_path] }
+    end
+  end
+
+  def present_pages(saved_pages)
+    saved_pages.map { |saved_page| present_page(saved_page) }
+  end
```

Comment:
I'm a bit wary of single-use methods... is this really better than

```ruby
  def index
    render_api_response(saved_pages: user.saved_pages.map { |page| present_page(page) })
  end
```

Or even

```ruby
  def index
    render_api_response(saved_pages: user.saved_pages.map(&:to_hash))
  end
```

Reply:
Fair point, it can be shrunk, let's roll that conversation into this one: https://github.com/alphagov/account-api/pull/69/files#r633509206